poly/src/polynomial/univariate/dense.rs

```diff
@@ -413,29 +413,37 @@ impl<'a, F: Field> AddAssign<(F, &'a DensePolynomial<F>)> for DensePolynomial<F>
 impl<'a, F: Field> AddAssign<&'a SparsePolynomial<F>> for DensePolynomial<F> {
     #[inline]
     fn add_assign(&mut self, other: &'a SparsePolynomial<F>) {
+        // No need to modify self if other is zero
+        if other.is_zero() {
+            return;
+        }
+
+        // If the first polynomial is zero, just copy the second one.
         if self.is_zero() {
-            self.coeffs.truncate(0);
+            self.coeffs.clear();
             self.coeffs.resize(other.degree() + 1, F::zero());
-
             for (i, coeff) in other.iter() {
                 self.coeffs[*i] = *coeff;
             }
-        } else if other.is_zero() {
         } else {
-            // If `other` has higher degree than `self`, create a dense vector
-            // storing the upper coefficients of the addition
-            let mut upper_coeffs = match other.degree() > self.degree() {
-                true => vec![F::zero(); other.degree() - self.degree()],
-                false => Vec::new(),
-            };
+            // If neither polynomial is zero, we proceed to add the terms.
+            let lhs_degree = self.degree();
+
+            // Resize the coefficients of the left-hand side if necessary.
+            // This is done to ensure that the left-hand side has enough coefficients.
+            let max_degree = lhs_degree.max(other.degree());
+            self.coeffs.resize(max_degree + 1, F::zero());
+
+            // Add the coefficients of the right-hand side to the left-hand side.
+            // - For pow <= lhs_degree, add the coefficients.
+            // - For pow > lhs_degree, set the coefficients (no addition is needed).
             for (pow, coeff) in other.iter() {
-                if *pow <= self.degree() {
+                if *pow <= lhs_degree {
                     self.coeffs[*pow] += coeff;
                 } else {
-                    upper_coeffs[*pow - self.degree() - 1] = *coeff;
+                    self.coeffs[*pow] = *coeff;
                 }
```

Comment:
We might need to truncate to remove a leading zero if the top coefficients are additive inverses. (The old code seems to have missed this edge case)

Reply:
fixed with a unit test to check this behaviour